contracts/mainnet/connectors/morpho-blue/helpers.sol

```diff
@@ -0,0 +1,190 @@
+                    _onBehalf
+                );
+
+                _amt = UtilsLib.min(_maxBalance, _amtDebt);
+            } else {
+                _amt = _maxBalance;
+            }
+        }
+
+        // Perform eth to weth conversion if necessary
+        convertEthToWeth(
+            true,
+            _isModeCollateral
+                ? TokenInterface(_marketParams.collateralToken)
+                : TokenInterface(_marketParams.loanToken),
+            _amt
+        );
+
+        return (_id, _marketParams, _amt);
+    }
+
+    /// @notice Handles Eth to Weth conversion if shares are provided.
+    function _performEthToWethSharesConversion(
+        MarketParams memory _marketParams,
+        uint256 _shares,
+        address _onBehalf,
+        uint256 _getId,
+        bool _isRepay
+    ) internal returns (Id _id, MarketParams memory, uint256 _assets) {
+        uint256 _shareAmt = getUint(_getId, _shares);
+        bool _isEth = _marketParams.loanToken == ethAddr;
+
+        _marketParams = updateTokenAddresses(_marketParams);
+
+        _id = _marketParams.id();
+
+        // Handle the max share case
+        if (_shares == type(uint256).max) {
+            uint256 _maxBalance = _isEth
+                ? address(this).balance
+                : TokenInterface(_marketParams.loanToken).balanceOf(
+                    address(this)
+                );
+
+            // If it's repay calculate the min of balance available and debt to repay
+            if (_isRepay) {
+                _assets = UtilsLib.min(
+                    _maxBalance,
+                    getPaybackBalance(_id, _marketParams, _onBehalf)
+                );
+            } else {
+                _assets = _maxBalance;
+            }
+        } else {
+            (
+                uint256 totalSupplyAssets,
+                uint256 totalSupplyShares,
+                uint256 totalBorrowAssets,
+                uint256 totalBorrowShares
+            ) = MORPHO_BLUE.expectedMarketBalances(_marketParams);
+
+            if (_isRepay) {
+                _assets = _shareAmt.toAssetsUp(
+                    totalBorrowAssets,
+                    totalBorrowShares
+                );
+            } else {
+                _assets = _shareAmt.toAssetsUp(
+                    totalSupplyAssets,
+                    totalSupplyShares
+                );
+            }
+        }
+
+        // Perform ETH to WETH conversion if necessary
+        convertEthToWeth(
+            true,
+            TokenInterface(_marketParams.loanToken),
+            _assets
+        );
+
+        return (_id, _marketParams, _assets);
+    }
+
+    /// @notice Returns the payback balance in assets.
+    function getPaybackBalance(
+        Id _id,
+        MarketParams memory _marketParams,
+        address _onBehalf
+    ) internal view returns (uint256 _assets) {
+        uint256 _borrowedShareAmt = MORPHO_BLUE
+            .position(_id, _onBehalf)
+            .borrowShares;
```

Comment:
For your information, you could be doing:

`using MorphoBalancesLib for IMorpho`
and save a SLOAD by calling `MORPHO_BLUE.borrowShares(_id, _onBehalf)`

Reply:
Agreed! Updated this as well